torch/nn/utils/spectral_norm.py

```diff
@@ -0,0 +1,112 @@
+"""
+Spectral Normalization from https://arxiv.org/abs/1802.05957
+"""
+import torch
+from torch.nn.functional import normalize
+from torch.nn.parameter import Parameter
+
+
+class SpectralNorm(object):
+
+    def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
+        self.name = name
+        self.n_power_iterations = n_power_iterations
+        self.eps = eps
+
+    def compute_weight(self, module):
+        weight = module._parameters[self.name + '_org']
+        u = module._buffers[self.name + '_u']
+        height = weight.size(0)
+        weight_mat = weight.view(height, -1)
+        for _ in range(self.n_power_iterations):
+            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
+            u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+        sigma = torch.dot(u, torch.matmul(weight_mat, v))
+        weight.data /= sigma
+        return weight, u
+
+    def remove(self, module):
+        weight = module._parameters[self.name + '_org']
+        del module._parameters[self.name]
+        del module._buffers[self.name + '_u']
+        del module._parameters[self.name + '_org']
+        module.register_parameter(self.name, weight)
+
+    def __call__(self, module, inputs):
+        weight, u = self.compute_weight(module)
+        setattr(module, self.name, weight)
+        setattr(module, self.name + '_u', u)
+
+    @staticmethod
+    def apply(module, name, n_power_iterations, eps):
+        fn = SpectralNorm(name, n_power_iterations, eps)
+        weight = module._parameters[name]
+        height = weight.size(0)
+
+        u = normalize(weight.data.new(height).normal_(0, 1), dim=0, eps=fn.eps)
```

Comment:
nit: `.data` is not needed, and please use `new_empty` instead of `new`.

Reply:
I didn't know that method, thank you.